### PR TITLE
Task: Set config

### LINKED ISSUE #3466: [ioctl] build config set command line into new ioctl
### What would you like to be added
 - Step 1: refactor  command  [_configSetCmd](https://github.com/iotexproject/iotex-core/blob/master/ioctl/config/configsetget.go#L62) into `./ioctl/newcmd/config/config_set.go`
 - Step 2: create `./ioctl/newcmd/config/config_set_test.go` to add unit test 

#### Notes:
1. Use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to construct the Cobra command.
2. [Output package](https://github.com/iotexproject/iotex-core/blob/master/ioctl/output) is deprecated. `e.g.` `output.PrintError` could be replaced with `errors.Wrap`.
3. Global variables (config.ReadConfig  config.DefaultConfigFile etc.) have been wrapped into client interface. Please use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to access them if needed
4. The logic of new code should be similar to [the old one](https://github.com/iotexproject/iotex-core/blob/master/ioctl/config/configsetget.go#L62).
5. In each unit test, the results should be checked whether they are as the expected. 

#### Example:
1. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountinfo.go
2. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountdelete.go
3. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountcreateadd.go

### Why is this needed
Unit tests couldn't be added for the old ioctl command (./ioctl/cmd)

### PR DESCRIPTION
**Pull request by pocockn**
# Description

Step 1: refactor command [_configSetCmd](https://github.com/iotexproject/iotex-core/blob/master/ioctl/config/configsetget.go#L62) into ./ioctl/newcmd/config/config_set.go
Step 2: create ./ioctl/newcmd/config/config_set_test.go to add unit test

Fixes #3466

## Type of change
Please delete options that are not relevant.
- [x] Code refactor or improvement

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test

### CODE REVIEW

File: ioctl/newcmd/config/config_set.go [hunk 1]
```diff
@@ -5,3 +5,62 @@
...
+	"github.com/spf13/cobra"
+
+	"github.com/iotexproject/iotex-core/ioctl"
+	"github.com/iotexproject/iotex-core/ioctl/config"
+)
+
+var (
+	_configSetUse = map[config.Language]string{
+		config.English: "set VARIABLE VALUE",
+		config.Chinese: "set 变量值",
```
Comment (huof6829) [inline ioctl/newcmd/config/config_set.go:23]: `变量  值`

File: ioctl/newcmd/config/config_set.go [hunk 2]
```diff
@@ -5,3 +5,62 @@
...
+	}
+	_configSetUseCmdShorts = map[config.Language]string{
+		config.English: "Set config fields for ioctl",
+		config.Chinese: "为 ioctl 设置配置字段",
+	}
+	_configSetUseCmdLong = map[config.Language]string{
+		config.English: "Set config fields for ioctl\nValid Variables: [" + strings.Join(_validGetArgs, ", ") + "]",
+		config.Chinese: "为 ioctl 设置配置字段\n有效变量: [" + strings.Join(_validGetArgs, ", ") + "]",
```
Comment (huof6829) [inline ioctl/newcmd/config/config_set.go:31]: _validArgs

File: ioctl/newcmd/config/config_set.go [hunk 3]
```diff
@@ -5,3 +5,62 @@
...
+	}
+
+	// Insecure represents the insecure connect option of grpc dial, default is false
+	Insecure = false
```
Comment (huof6829) [inline ioctl/newcmd/config/config_set.go:35]: delete.
It's used for insecure flag. You can use  `client.SetInsecureWithFlag`

File: ioctl/newcmd/config/config_set.go [hunk 4]
```diff
@@ -5,3 +5,62 @@
...
+func NewConfigSetCmd(client ioctl.Client) *cobra.Command {
+	use, _ := client.SelectTranslation(_configSetUse)
+	short, _ := client.SelectTranslation(_configSetUseCmdShorts)
+	long, _ := client.SelectTranslation(_configSetUseCmdLong)
+
+	return &cobra.Command{
+		Use:       use,
+		Short:     short,
+		Long:      long,
+		ValidArgs: _validGetArgs,
```
Comment (huof6829) [inline ioctl/newcmd/config/config_set.go:48]: _validArgs

File: ioctl/newcmd/config/config.go [hunk 5]
```diff
@@ -8,6 +8,8 @@ package config
 
 import (
 	"fmt"
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
```
[the remark this answers is not in the document]
Comment (pocockn) [reply ioctl/newcmd/config/config.go]: @huof6829  Thanks, done 👍

**Comment by pocockn** [comment]
@huof6829 I've made those changes. 

I'm having some issues running the application using `make run` 

I get the following:

```
{"level":"error","ts":1658399150.6483269,"caller":"factory/statedb.go:146","msg":"Failed to execute state factory creation option 0x101341560: failed to open kvstore patch, /var/data/trie.db.patch: open /var/data/trie.db.patch: no such file or directory","ioAddr":"io19qjnrl0649t0z2jgedz8unex3emc9kqghzvefj"}
```

I've followed the instructions here: https://github.com/iotexproject/iotex-bootstrap#join-mainnet but still get the error.

Is there anything else I need to do? Do you have a discord to chat on or something? 

Thanks!

**Comment by huof6829** [comment]
> https://github.com/iotexproject/iotex-bootstrap#join-mainnet

```
mkdir -p /var/data
curl https://raw.githubusercontent.com/iotexproject/iotex-bootstrap/v1.6.3/trie.db.patch > /var/data/trie.db.patch
```
/var/data/trie.db.patch should be existed and then `make run`

File: ioctl/newcmd/config/config_set.go [hunk 6]
```diff
@@ -5,3 +5,62 @@
...
+	use, _ := client.SelectTranslation(_configSetUse)
+	short, _ := client.SelectTranslation(_configSetUseCmdShorts)
+	long, _ := client.SelectTranslation(_configSetUseCmdLong)
+
+	cmd := &cobra.Command{
+		Use:       use,
+		Short:     short,
+		Long:      long,
+		ValidArgs: _validArgs,
+		Args: func(cmd *cobra.Command, args []string) error {
```
Comment (huof6829) [inline ioctl/newcmd/config/config_set.go:46]: use cobra.ExactValidArgs.  
add more test datas to test it

File: ioctl/newcmd/config/config.go [hunk 7]
```diff
@@ -128,6 +131,78 @@ func (c *info) reset() error {
...
+		err2 := validator.ValidateAddress(args[1])
+		if err1 != nil && err2 != nil {
+			return "", errors.New(fmt.Sprintf("failed to validate alias or address %s", args[1]))
+		}
+		c.readConfig.DefaultAccount.AddressOrAlias = args[1]
+	case "language":
+		lang := isSupportedLanguage(args[1])
+		if lang == -1 {
+			return "", errors.New(
+				fmt.Sprintf("Language %s is not supported\nSupported languages: %s",
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:184]: errors.Errorf

File: ioctl/newcmd/config/config.go [hunk 8]
```diff
@@ -128,6 +131,78 @@ func (c *info) reset() error {
...
+					args[1], _supportedLanguage))
+		}
+		c.readConfig.Language = _supportedLanguage[lang]
+	case "nsv2height":
+		height, err := strconv.ParseUint(args[1], 10, 64)
+		if err != nil {
+			return "", errors.New(fmt.Sprintf("invalid height %d", height))
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:191]: errors.Errorf

File: ioctl/newcmd/config/config.go [hunk 9]
```diff
@@ -128,6 +131,78 @@ func (c *info) reset() error {
...
+		default:
+			return "", errors.New(
+				fmt.Sprintf("Explorer %s is not valid\nValid explorers: %s",
+					args[1], append(_validExpl, "custom")))
+		}
+	case "defaultacc":
+		err1 := validator.ValidateAlias(args[1])
+		err2 := validator.ValidateAddress(args[1])
+		if err1 != nil && err2 != nil {
+			return "", errors.New(fmt.Sprintf("failed to validate alias or address %s", args[1]))
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:177]: errors.Errorf

File: ioctl/newcmd/config/config.go [hunk 10]
```diff
@@ -128,6 +131,78 @@ func (c *info) reset() error {
...
+		case args[1] == "custom":
+			fmt.Println(`Please enter a custom link below:("Example: iotexscan.io/action/")`)
+			var link string
+			_, err := fmt.Scanln(&link)
+			if err != nil {
+				return "", err
+			}
+			match, err := regexp.MatchString(_urlPattern, link)
+			if err != nil {
+				return "", errors.New(fmt.Sprintf("failed to validate link %s", link))
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:161]: errors.Errorf

File: ioctl/newcmd/config/config.go [hunk 11]
```diff
@@ -128,6 +131,78 @@ func (c *info) reset() error {
...
+			}
+			if match {
+				c.readConfig.Explorer = link
+			} else {
+				return "", errors.New(fmt.Sprintf("invalid link %s", link))
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:166]: errors.Errorf

File: ioctl/newcmd/config/config.go [hunk 12]
```diff
@@ -128,6 +131,78 @@ func (c *info) reset() error {
 	return nil
 }
 
+// set sets config variable
+func (c *info) set(args []string, insecure bool) (string, error) {
+	switch args[0] {
+	case "endpoint":
+		if !isValidEndpoint(args[1]) {
+			return "", errors.New(fmt.Sprintf("endpoint %s is not valid", args[1]))
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:139]: errors.Errorf

File: ioctl/newcmd/config/config.go [hunk 13]
```diff
@@ -128,6 +131,78 @@ func (c *info) reset() error {
...
+		c.readConfig.Wallet = args[1]
+	case "explorer":
+		lowArg := strings.ToLower(args[1])
+		switch {
+		case isValidExplorer(lowArg):
+			c.readConfig.Explorer = lowArg
+		case args[1] == "custom":
+			fmt.Println(`Please enter a custom link below:("Example: iotexscan.io/action/")`)
+			var link string
+			_, err := fmt.Scanln(&link)
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:155]: These should be put into `client` (refer to `AskToConfirm`).  call `client.func` here.   we can mock the new function in test, so the left code can be run.

File: ioctl/newcmd/config/config.go [hunk 14]
```diff
@@ -128,6 +131,78 @@ func (c *info) reset() error {
...
+			}
+		default:
+			return "", errors.New(
+				fmt.Sprintf("Explorer %s is not valid\nValid explorers: %s",
+					args[1], append(_validExpl, "custom")))
+		}
+	case "defaultacc":
+		err1 := validator.ValidateAlias(args[1])
+		err2 := validator.ValidateAddress(args[1])
+		if err1 != nil && err2 != nil {
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:176]: if ;err==nil{
else if  ;err==nil{
else {
can be better

**Comment by pocockn** [comment]
> > https://github.com/iotexproject/iotex-bootstrap#join-mainnet
> 
> ```
> mkdir -p /var/data
> curl https://raw.githubusercontent.com/iotexproject/iotex-bootstrap/v1.8.1/trie.db.patch > /var/data/trie.db.patch
> ```
> 
> /var/data/trie.db.patch should be existed and then `make run`

Awesome, that worked. Thanks

File: ioctl/newcmd/config/config.go [hunk 15]
```diff
@@ -128,6 +131,78 @@ func (c *info) reset() error {
 	return nil
 }
 
+// set sets config variable
+func (c *info) set(args []string, insecure bool) (string, error) {
+	switch args[0] {
+	case "endpoint":
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:137]: the var is used in get and set function.  const  var is better, for hand write by mistake

File: ioctl/newcmd/config/config.go [hunk 16]
```diff
@@ -128,6 +131,76 @@ func (c *info) reset() error {
...
+	case "analyserEndpoint":
+		c.readConfig.AnalyserEndpoint = args[1]
+	case "wallet":
+		c.readConfig.Wallet = args[1]
+	case "explorer":
+		lowArg := strings.ToLower(args[1])
+		switch {
+		case isValidExplorer(lowArg):
+			c.readConfig.Explorer = lowArg
+		case args[1] == "custom":
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:152]: `"custom"` can add description in `_configSetUseCmdLong`

File: ioctl/newcmd/config/config_set.go (hunk 6 above)
Comment (pocockn) [reply ioctl/newcmd/config/config_set.go:46]: I'm not sure we can use `cobra.ExactValidArgs` as the user can enter custom args after the config field they are setting. For example `"nsv2height", "44"` which gives the error `invalid argument "44" for "config"` as 44 isn't in the list of valid args.

We could use `cobra.ExactArgs` which checks the number of args?

Thanks

File: ioctl/newcmd/config/config.go (hunk 16 above)
Comment (pocockn) [reply ioctl/newcmd/config/config.go:153]: Added that but the Chinese translation may need tweaking 😬

File: ioctl/newcmd/config/config.go (hunk 13 above)
Comment (pocockn) [reply ioctl/newcmd/config/config.go]: Added and passed the client into the function, let me know if that's the solution you had in mind.

File: ioctl/newcmd/config/config.go [hunk 17]
```diff
@@ -128,10 +132,70 @@ func (c *info) reset() error {
 	return nil
 }
 
+// set sets config variable
+func (c *info) set(args []string, insecure bool, client ioctl.Client) (string, error) {
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:136]: delete `insecure bool`
use client function in line 143

File: ioctl/newcmd/config/config.go (hunk 16 above)
Comment (huof6829) [reply ioctl/newcmd/config/config.go:153]: good.

File: ioctl/validator/validator.go [hunk 18]
```diff
@@ -24,9 +24,9 @@ var (
 	ErrNonPositiveNumber = errors.New("invalid number that is not positive")
 	// ErrInvalidStakeDuration indicates error for invalid stake duration
 	ErrInvalidStakeDuration = errors.New("stake duration must be within 0 and 1050 and in multiples of 7")
-	// ErrInvalidCandidateName indicates error for invalid candidate name
+	// ErrLongCandidateName ErrInvalidCandidateName indicates error for invalid candidate name
 	ErrLongCandidateName = errors.New("invalid length of candidate name that is more than 12 ")
-	// ErrInvalidCandidateName indicates error for invalid candidate name (for ioctl stake2 command)
+	// ErrStake2CandidateName ErrInvalidCandidateName indicates error for invalid candidate name (for ioctl stake2 command)
```
Comment (huof6829) [inline ioctl/validator/validator.go:29]: @pocockn  good finding! Thx

File: ioctl/client.go [hunk 19]
```diff
@@ -183,6 +192,26 @@ func (c *client) AskToConfirm(info string) (bool, error) {
 	return strings.EqualFold(confirm, "yes"), nil
 }
 
+// CustomLink is used when setting the explorer config field with a custom link.
+func (c *client) CustomLink() (string, error) {
+	var link string
+	_, err := fmt.Scanln(&link)
+	if err != nil {
```
Comment (huof6829) [inline ioctl/client.go:199]: in one line  if _, err := ; err!=nil{

File: ioctl/client.go [hunk 20]
```diff
@@ -183,6 +192,26 @@ func (c *client) AskToConfirm(info string) (bool, error) {
 	return strings.EqualFold(confirm, "yes"), nil
 }
 
+// CustomLink is used when setting the explorer config field with a custom link.
+func (c *client) CustomLink() (string, error) {
```
Comment (huof6829) [inline ioctl/client.go:196]: add  `// notest` at end of this line,  so coverage don't scan this function.

File: ioctl/newcmd/config/config.go [hunk 21]
```diff
@@ -128,10 +132,70 @@ func (c *info) reset() error {
...
+	switch args[0] {
+	case _endpoint:
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:138]: other case string constants  can also make to const var

File: ioctl/newcmd/config/config.go [hunk 22]
```diff
@@ -128,10 +132,70 @@ func (c *info) reset() error {
...
+		c.readConfig.Wallet = args[1]
+	case "explorer":
+		lowArg := strings.ToLower(args[1])
+		switch {
+		case isValidExplorer(lowArg):
+			c.readConfig.Explorer = lowArg
+		case args[1] == "custom":
+			link, err := client.CustomLink()
+			if err != nil {
+				return "", errors.Errorf("invalid link %s", link)
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:156]: errors.Wrapf(err,

File: ioctl/newcmd/config/config.go [hunk 23]
```diff
@@ -128,10 +132,70 @@ func (c *info) reset() error {
...
+		lang := isSupportedLanguage(args[1])
+		if lang == -1 {
+			return "", errors.Errorf("language %s is not supported\nSupported languages: %s",
+				args[1], _supportedLanguage)
+		}
+		c.readConfig.Language = _supportedLanguage[lang]
+	case "nsv2height":
+		height, err := strconv.ParseUint(args[1], 10, 64)
+		if err != nil {
+			return "", errors.Errorf("invalid height %d", height)
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:180]: errors.Wrapf(err,

File: ioctl/newcmd/config/config_test.go [hunk 24]
```diff
@@ -130,3 +130,79 @@ func TestConfigReset(t *testing.T) {
...
+		},
+		{
+			[]string{"unknownField", ""},
+			"no matching config",
+		},
+	}
+
+	for _, tc := range tcs {
+		setResult, err := info.set(tc.args, false, nil)
+		t.Logf("running %s \n result %s", tc, setResult)
```
Comment (huof6829) [inline ioctl/newcmd/config/config_test.go:201]: delete line201

File: ioctl/newcmd/config/config_set.go (hunk 6 above)
Comment (huof6829) [reply ioctl/newcmd/config/config_set.go:46]: yes，`cobra.ExactArgs` can replace check len(args)

File: ioctl/newcmd/config/config.go [hunk 25]
```diff
@@ -128,10 +132,70 @@ func (c *info) reset() error {
...
+			}
+			c.readConfig.Explorer = link
+		default:
+			return "", errors.Errorf("explorer %s is not valid\nValid explorers: %s",
+				args[1], append(_validExpl, "custom"))
+		}
+	case "defaultacc":
+		if err := validator.ValidateAlias(args[1]); err == nil {
+		} else if err = validator.ValidateAddress(args[1]); err == nil {
+			c.readConfig.DefaultAccount.AddressOrAlias = args[1]
```
Comment (huof6829) [inline ioctl/newcmd/config/config.go:166]: move to line 170

**Review by huof6829 — APPROVED** [review]

File: ioctl/newcmd/config/config_set.go (hunk 6 above)
Comment (Liuhaai) [inline ioctl/newcmd/config/config_set.go:46]: `cobra.ExactArgs(2)` ?

**Review by Liuhaai — APPROVED** [review]

**Comment by Liuhaai** [comment]
good job

**Pull request merged**